src/components/cl/hier/allreduce/allreduce_rab.c

```diff
@@ -143,18 +146,33 @@ ucc_cl_hier_allreduce_rab_init_schedule(ucc_base_coll_args_t *coll_args,
         n_tasks++;
     }
 
-    UCC_CHECK_GOTO(ucc_event_manager_subscribe(
-                       &schedule->super, UCC_EVENT_SCHEDULE_STARTED, tasks[0],
-                       ucc_task_start_handler),
-                   out, status);
-
+    /* subscription logic is different depending on top level schedule type
+     * being used
+     */
     UCC_CHECK_GOTO(ucc_schedule_add_task(schedule, tasks[0]), out, status);
-    for (i = 1; i < n_tasks; i++) {
-        UCC_CHECK_GOTO(
-            ucc_event_manager_subscribe(tasks[i - 1], UCC_EVENT_COMPLETED,
-                                        tasks[i], ucc_task_start_handler),
-            out, status);
-        UCC_CHECK_GOTO(ucc_schedule_add_task(schedule, tasks[i]), out, status);
+    if (n_frags > 1) {
+        UCC_CHECK_GOTO(ucc_task_subscribe_dep(&schedule->super, tasks[0],
+                                            UCC_EVENT_SCHEDULE_STARTED),
+                    out, status);
+        for (i = 1; i < n_tasks; i++) {
+            UCC_CHECK_GOTO(ucc_schedule_add_task(schedule, tasks[i]), out, status);
+            UCC_CHECK_GOTO(ucc_task_subscribe_dep(tasks[i-1], tasks[i],
```

Comment:
it's what i did initially, but it doesn't work for persistent collectives because of deps counter

Reply:
we start collective using schedule_pipelined_post, right? doesn't it reset deps_satisfied to 0? Why exactly dep counter is broken for persistent?

Reply:
it does, but we use schedule pipelined init only for pipelined schedule otherwise it will be ucc_schedule_start

Reply:
Yes, but this is what we call: ucc_cl_hier_rab_allreduce_start->schedule_pipelined_post. Am i missing smth?
Also even if we call ucc_schedule_start. I think it would be cleaner to add n_deps_satisfied = 0 to ucc_schedule_start and then still use "subscribe_dep" alone.

Reply:
as discussed over phone we can't do it because it will break pipelined schedules. Action items  for next PRs would be to 
1. check if bcast 2 step has similar bug
2. check if use pipelined schedule only will not harm performance